### PR TITLE
feat: downgrade logging level when AppMap recording is enabled

### PR DESCRIPTION
**Pull request by symwell**
Don't show noise in the console.

- Show a warning when the agent does something I'm not necessarily expecting, or when I've made a mistake. So, if set `APPMAP=true`, or `APPMAP=false`, or one of the `APPMAP_RECORD_` variables is set to true or false, log at `info`.
- If any of those variables is set to a value other than `true` or `false`, write a message at `warning`. This is new behavior.
- Make the comparison with `true` or `false` be case-insensitive. Before it was done with `== "true"` and `== "false"`.
- If there's an environment variable that starts with the prefix `APPMAP_RECORD_` but doesn't match one of the supported recording types, log a message at `warning`.  This is also new behavior.
- We've gone back and forth about what level to use when recording by default. On balance, we should log this message at `info`: it's presumably exactly the behavior they were hoping for when they installed the agent, and they'll see all the AppMaps that got generated.

Fixes https://github.com/getappmap/appmap-python/issues/182.

Tested with `poetry run pytest -s -v appmap/test/test_django.py`.

### CODE REVIEW

**Comment by apotterri** [comment]
Can you add something (Loom, asciinema, copy-and-paste terminal session, whatever) that shows the output for each of the new cases, please?

**Comment by symwell** [comment]
I would like to check this with testcases.

I changed the testcases to assert the logger output.  I found that WARNING output gets produced as expected but INFO output isn't.  I thought the logger was configured to not show INFO output but after changing that in both testcases and in `detect_enabled.py` I don't see any INFO output even though code execution gets to the `logger.info` statement.

**Comment by apotterri** [comment]
>I would like to check this with testcases.

I agree that it's important to verify it with testcases. But, I'd also like to see what the user experience is actually going to look like, as seen in a terminal.

**Comment by symwell** [comment]
1) APPMAP=true, `master` (before)
```
PYTHONPATH=/home/test/src/appmap-python APPMAP=true python3 -m appmap.unittest program_test.py
[2022-10-20 13:07:44,163] WARNING appmap._implementation.detect_enabled: AppMap recording is enabled because APPMAP=true
WARNING:appmap._implementation.detect_enabled:AppMap recording is enabled because APPMAP=true
.
----------------------------------------------------------------------
Ran 1 test in 0.023s

OK
```
APPMAP=true, `sw/dont_show_warning_when_appmap_is_enabled` (after)
```
PYTHONPATH=/home/test/src/appmap-python APPMAP=true python3 -m appmap.unittest program_test.py
.
----------------------------------------------------------------------
Ran 1 test in 0.023s

OK
```

2) APPMAP=false, `master` (before)
```
PYTHONPATH=/home/test/src/appmap-python APPMAP=false python3 -m appmap.unittest program_test.py
WARNING:__main__:AppMap disabled. Did you forget to set APPMAP=true?
[2022-10-20 13:10:29,134] WARNING appmap.unittest: AppMap disabled. Did you forget to set APPMAP=true?
WARNING:appmap.unittest:AppMap disabled. Did you forget to set APPMAP=true?
.
----------------------------------------------------------------------
Ran 1 test in 0.000s

OK
```
APPMAP=false, `sw/dont_show_warning_when_appmap_is_enabled` (after)
```
PYTHONPATH=/home/test/src/appmap-python APPMAP=false python3 -m appmap.unittest program_test.py
.
----------------------------------------------------------------------
Ran 1 test in 0.000s

OK
```

3) APPMAP=invalid, `master` (before)
```
PYTHONPATH=/home/test/src/appmap-python APPMAP=invalid python3 -m appmap.unittest program_test.py
[2022-10-20 13:46:11,338] WARNING appmap._implementation.detect_enabled: AppMap recording is enabled because will record by default
WARNING:appmap._implementation.detect_enabled:AppMap recording is enabled because will record by default
.
----------------------------------------------------------------------
Ran 1 test in 0.023s

OK
```
APPMAP=invalid, `sw/dont_show_warning_when_appmap_is_enabled` (after)
```
PYTHONPATH=/home/test/src/appmap-python APPMAP=invalid python3 -m appmap.unittest program_test.py
AppMap recording is not enabled because APPMAP=invalid is an invalid option.
AppMap recording is not enabled because APPMAP=invalid is an invalid option.
AppMap recording is not enabled because APPMAP=invalid is an invalid option.
AppMap recording is not enabled because APPMAP=invalid is an invalid option.
[2022-10-20 13:43:02,650] WARNING appmap._implementation.detect_enabled: AppMap recording is not enabled because APPMAP=invalid is an invalid option.
[2022-10-20 13:43:02,650] WARNING appmap._implementation.detect_enabled: AppMap recording is not enabled because APPMAP=invalid is an invalid option.
[2022-10-20 13:43:02,650] WARNING appmap._implementation.detect_enabled: AppMap recording is not enabled because APPMAP=invalid is an invalid option.
[2022-10-20 13:43:02,650] WARNING appmap._implementation.detect_enabled: AppMap recording is not enabled because APPMAP=invalid is an invalid option.
[2022-10-20 13:43:02,753] WARNING appmap._implementation.detect_enabled: AppMap recording is not enabled because APPMAP=invalid is an invalid option.
WARNING:appmap._implementation.detect_enabled:AppMap recording is not enabled because APPMAP=invalid is an invalid option.
.
----------------------------------------------------------------------
Ran 1 test in 0.000s

OK
```

4) APPMAP_RECORD_INVALID=true, `master` (before)
```
PYTHONPATH=/home/test/src/appmap-python APPMAP_RECORD_INVALID=true python3 -m appmap.unittest program_test.py
[2022-10-20 13:48:59,467] WARNING appmap._implementation.detect_enabled: AppMap recording is enabled because will record by default
WARNING:appmap._implementation.detect_enabled:AppMap recording is enabled because will record by default
.
----------------------------------------------------------------------
Ran 1 test in 0.023s

OK
```
APPMAP_RECORD_INVALID=true, `sw/dont_show_warning_when_appmap_is_enabled` (after)
```
PYTHONPATH=/home/test/src/appmap-python APPMAP_RECORD_INVALID=true python3 -m appmap.unittest program_test.py
AppMap recording is not enabled because APPMAP_RECORD_INVALID is an invalid recording method.
AppMap recording is not enabled because APPMAP_RECORD_INVALID is an invalid recording method.
AppMap recording is not enabled because APPMAP_RECORD_INVALID is an invalid recording method.
AppMap recording is not enabled because APPMAP_RECORD_INVALID is an invalid recording method.
[2022-10-20 13:49:24,044] WARNING appmap._implementation.detect_enabled: AppMap recording is not enabled because APPMAP_RECORD_INVALID is an invalid recording method.
[2022-10-20 13:49:24,044] WARNING appmap._implementation.detect_enabled: AppMap recording is not enabled because APPMAP_RECORD_INVALID is an invalid recording method.
[2022-10-20 13:49:24,044] WARNING appmap._implementation.detect_enabled: AppMap recording is not enabled because APPMAP_RECORD_INVALID is an invalid recording method.
[2022-10-20 13:49:24,044] WARNING appmap._implementation.detect_enabled: AppMap recording is not enabled because APPMAP_RECORD_INVALID is an invalid recording method.
[2022-10-20 13:49:24,147] WARNING appmap._implementation.detect_enabled: AppMap recording is not enabled because APPMAP_RECORD_INVALID is an invalid recording method.
WARNING:appmap._implementation.detect_enabled:AppMap recording is not enabled because APPMAP_RECORD_INVALID is an invalid recording method.
.
----------------------------------------------------------------------
Ran 1 test in 0.000s

OK
```

**Comment by apotterri** [comment]
It doesn't seem ideal to print the same message multiple times. Can we do something about that?

**Comment by symwell** [comment]
Screenshot diffs.
Top is `master`
Bottom is `sw/dont_show_warning_when_appmap_is_enabled`

1) APPMAP=true

v2_APPMAP_true


2) APPMAP=false

v2_APPMAP_false


3) APPMAP=invalid

v2_APPMAP_invalid


4) APPMAP_RECORD_INVALID=true

v2_APPMAP_RECORD_INVALID


5) APPMAP_RECORD_REQUESTS=true

v2_APPMAP_RECORD_REQUESTS_true


6) APPMAP_RECORD_REQUESTS=false


v2_APPMAP_RECORD_REQUESTS_false

**Comment by symwell** [comment]
`APPMAP_RECORD_UNITTEST` shows logs on stdout without the `WARNING`.

```
PYTHONPATH=/home/test/src/appmap-python APPMAP_RECORD_UNITTEST=invalid python3 -m appmap.unittest program_test.py
[2022-10-21 13:04:24,539] WARNING appmap._implementation.detect_enabled: AppMap recording is not enabled because APPMAP_RECORD_UNITTEST=invalid is an invalid option.
WARNING:appmap._implementation.detect_enabled:AppMap recording is not enabled because APPMAP_RECORD_UNITTEST=invalid is an invalid option.
.
----------------------------------------------------------------------
Ran 1 test in 0.000s

OK
```

File: appmap/_implementation/configuration.py [hunk 1]
```diff
@@ -383,9 +383,7 @@ def __init__(self, *args, **kwargs):
 
 
 def initialize():
+    DetectEnabled.initialize()
     Config().initialize()
     Importer.use_filter(BuiltinFilter)
     Importer.use_filter(ConfigFilter)
-
-
-initialize()
```
Comment (apotterri) [inline appmap/_implementation/configuration.py:391]: Why did you remove this call to `initialize`()?

Comment (symwell) [reply appmap/_implementation/configuration.py:391]: [Removing](https://github.com/getappmap/appmap-python/pull/183/commits/edf1f514fb0f96ab25a12ff7ae3619dbf740277c) it lead to not printing 4 WARNINGs unnecessarily.

I've since found a different way to not print these warnings. I can add this call back.

The tests pass without this call.

Comment (apotterri) [reply appmap/_implementation/configuration.py:391]: Yes, please it put back, for the sake of consistency with other modules.

**Comment by symwell** [comment]
Created separate [issue ](https://github.com/getappmap/appmap-python/issues/185) to address configuring logging for `APPMAP_RECORD_UNITTEST`.

**Review by apotterri — CHANGES_REQUESTED** [review]
Please fix the body of the comment on the comment so it describes the change you made.

**Review by apotterri — APPROVED** [review]
Looks good, thanks for making this update.
